### ISSUE TITLE
Using closestKodein as Delegate for KodeinAware.kodein Causes NPE

### ISSUE THREAD

**Issue by eygraber** [Kodein-Framework/Kodein-DI#142]
The following code crashes when the `Service` is created:

```
class FCMService : FirebaseMessagingService(), KodeinAware {
  override val kodein by closestKodein()
}
```

Changing it to the following fixes it:

```
class FCMService : FirebaseMessagingService(), KodeinAware {
  private val _kodein: Kodein by closestKodein()

  override val kodein = Kodein.lazy {
    extend(_kodein)
  }
}
```

Stacktrace is:

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.myapp, PID: 5290
    java.lang.RuntimeException: Unable to instantiate service com.myapp.notifications.FCMService: java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.Context android.content.Context.getApplicationContext()' on a null object reference
        at android.app.ActivityThread.handleCreateService(ActivityThread.java:3324)
        at android.app.ActivityThread.-wrap4(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1677)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.Context android.content.Context.getApplicationContext()' on a null object reference
        at android.content.ContextWrapper.getApplicationContext(ContextWrapper.java:109)
        at org.kodein.di.android.ClosestKt.closestKodein(closest.kt:23)
        at org.kodein.di.android.ClosestKt.access$closestKodein(closest.kt:1)
        at org.kodein.di.android.ClosestKodeinInContextPropertyDelegateProvider$provideDelegate$1.invoke(closest.kt:37)
        at org.kodein.di.android.ClosestKodeinInContextPropertyDelegateProvider$provideDelegate$1.invoke(closest.kt:36)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at com.myapp.notifications.FCMService.getKodein(Unknown Source:7)
        at org.kodein.di.KodeinWrapper.<init>(KodeinAware.kt:205)
        at org.kodein.di.KodeinAwareKt.On(KodeinAware.kt:219)
        at com.myapp.notifications.FCMService.<init>(FCMService.kt:66)
        at java.lang.Class.newInstance(Native Method)
        at android.app.ActivityThread.handleCreateService(ActivityThread.java:3321)
        at android.app.ActivityThread.-wrap4(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1677) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6494) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807) 
```

**Comment by BulatMukhutdinov** [comment]
any progress on this?

**Comment by SalomonBrys** [comment]
I suppose this is not the entire code...
In your stacktrace, there's I need to understand this line:

```
at com.myapp.notifications.FCMService.<init>(FCMService.kt:66)
```

can you provide it ?

**Comment by jackqack** [comment]
Same problem with this following example from [Kodein DI on Android](http://kodein.org/Kodein-DI/?5.0/android#_component_based_kodein) page:
```
class MyActivity : Activity(), KodeinAware {
    private val _parentKodein by closestKodein() 
    override val kodein: Kodein = Kodein {
        extend(_parentKodein) 
        /* activity specific bindings */
    }
}
```

But it also fixes by initializing `kodein` object using `Kodein.lazy` or `by retainedKodein`.

**Comment by SalomonBrys** [comment]
@eygraber I can't reproduce this issue.
Do you have a sample project that I can checkout ?

**Comment by yaneq6** [comment]
I have same error when trying to mix `closestKodein()` with `... by on(context).instance()` . My case is not exactly same but simmilar to example below:
```
class MyActivity : Activity(), KodeinAware {
    private val _parentKodein by closestKodein() 
    override val kodein: Kodein = retainedKodein {
        extend(_parentKodein) 
        import(mainModule)
    }
    private val ui: MainUI by on(this).instance()
}
```
when mainModule implementation is something like:
```
val mainModule = KodeinModule("mainModule") {
    bind<MainUI>() with scoped(ActivityRetainedScope).singleton {
        MainUI(context, instance())
    }
}
```
The core of the problem, I guess is `on(this).instance()` because without `on(this)`  (and without scope on MainUI binding in mainModule, of course) the error not occurs.
It's looks like `on(this)` is trying to access application context before it will be set on activity, and it cause NPE simply, I guess...

**Comment by SalomonBrys** [comment]
After thinking about it (and doing 80% of the work implementing it), I've decided against it.
In Kodein, everything is lazy by default, **in retrieval**.

However, running the Kodein configuration block is kinda slow (because of all type extraction that happens, this can be dramatically improved by using erased), but that's OK as long as it is in application startup.
Having the configuration block running lazily by default would lead to weird cases: application startup would improve, but the first retrieval would be particularly slow.

I'll update the doc for this particular use case.

**Comment by SalomonBrys** [comment]
Already fixed in documentation 5.2: http://kodein.org/Kodein-DI/?5.2/android#_component_based_kodein

I really need to find a way to mark a documentation deprecated...
